src/scenic/core/scenarios.py

```diff
@@ -337,27 +334,47 @@ def generate(self, maxIterations=2000, verbosity=0, feedback=None):
                 rejection = e
                 continue
             rejection = None
-            ego = sample[self.egoObject]
 
-            # Normalize types of some built-in properties
-            for obj in objects:
+            # Ensure nothing else is lazy
+            for obj in self.objects:
                 sampledObj = sample[obj]
                 assert not needsSampling(sampledObj)
 
-            sampledObjects = [sample[obj] for obj in objects]
+            # Check validity of sample
+            rejection = self.checkSample(sample, activeReqs)
+
+            if rejection is None:
+                # The sample is valid, we're done
+                break
```

Comment:
Since the loop condition is `rejection is not None`, I think it would be clearer if you didn't use `break` but replaced this whole if-else with a simple
```python
if rejection is not None:
    optionallyDebugRejection()
```
Then there's only 1 way to exit the loop normally.